core/app/services/spree/add_item_to_cart.rb

```diff
@@ -0,0 +1,37 @@
+module Spree
+  class AddItemToCart
+    prepend Spree::ServiceModule::Base
+
+    def call(order:, variant:, quantity: nil, options: {})
```

Comment:
`params = {}` instead of named keyword arguments?

Reply:
Any advantages of that?

Reply:
Yeah, I don't really want something like this to happen
https://github.com/google/google-api-ruby-client/blob/1476c31b2f5c3e37e045c6ff774db4d301dd54db/generated/google/apis/calendar_v3/service.rb
Also - `params` seem more compact & such approach doesn't break Sandi Metz's rule about the max amount of arguments passed to the method.

Reply:
but then, we're losing way to make some params required(or we will have to handle it by ourselves). With keywords arguments we have `fail-fast` which is really good for debugging and fixing stuff. Also, doing something like `params[:order]` `params[:variant]` `params[:quantity]` doesn't look too good in code. 

I tried to use params hash few times with but ended up really messy and I'd like to keep keyword arguments.

Also, passing 5 params in a hash still breaks rule of the max amount of arguments. Moving them into hash doesn't change anything in this case :/

Reply:
`ActiveModel::Validations?` please no.

No variables in initializer please. Initializer is only for setting dependencies(which we don't have atm, but it will change).

We cannot avoid "ridiculous code" all the time. Most of services won't probably have 5+ arguments so it won't be so bad. Spree is quite big project and it's always tradeoff. ridiculosity vs readability. I prefer second one, since there is more reading code that writing it and we have to be explicit about what we want to do, since whole community will be using it too.

Reply:
> ActiveModel::Validations? please no

That was just a suggestion. We can use `dry-validations`, `hanami/validations` or whatever you prefer.
Keyword arguments are quite nice but I don't think that we should delegate input validation to them, this is kinda wrong. Ideally we should move validations to a separate layer completely.

> No variables in initializer please. Initializer is only for setting dependencies(which we don't have atm, but it will change).

What is the reason behind it? I believe it makes sense since we're basically using Services that implement Command Pattern therefore having an internal state they rely on. Basically instead of 
```
def call(order:, variant:, something:)
```
we'll have
```
def initialize(params = {})
  @order = params[:order]
end

def call
...
end

private

attr_reader :order
```
Technically speaking we're dependant on input therefore `order`, `variant` and other params are dependencies we're relying on. 
Am I missing something important here?

Reply:
```
That was just a suggestion. We can use dry-validations, hanami/validations or whatever you prefer.
```
Do we want to remove basic features of ruby and depend on external gem to replace them? It's not about validation, it's about passing correct data so method can be executed.
 

```
What is the reason behind it? I believe it makes sense since we're basically using Services that implement Command Pattern therefore having an internal state they rely on. Basically instead of
```
1) We don't want to rely on state at all

2) If we put params in initializer, where do we put dependencies? Dependency is not argument. Please read about DI https://en.wikipedia.org/wiki/Dependency_injection . That's the only reasonable way to deal with things like this and whole programming community is doing it. Noone invented anything better

3) We would have to write `@order = params[:order]` for each argument and then someone using it has no idea how it would be handled by class so we're introducing magic which we have already a lot. 

Why do you want to put everything in initializer? What's the advantage of that? It's only messing up debugging and flow of data. I can't see anything good in that solution.

Reply:
If we put everything(arguments and dependencies) in initializer, we won't be able to distinguish them. "Unfortunately" we're not using java and our compiler won't make work for us.

We don't have to keep everything in `call` method. That's why there is keeping state underneath between methods/services calls by `run`. Result of first `run` call is passed as params to second `run` method and then result of second to the next one, and so on. 

It kinda forces you to do better design since if you think that you "REALLY REALLY need state across methods" then it means that something is wrong with whole idea. Maybe service becomes too big or maybe it breaks SRP. 

Why do you think that classes are meant to have state? It's kind of thing that we're taught at universities about OOP but that's nothing about OOP.

Reply:
> If we put everything(arguments and dependencies) in initializer, we won't be able to distinguish them
To be perfectly honest I don't see a problem with it. 
It all comes down to the implementation - you can make everything perfectly distinguishable with initializers as well.

> Result of first run call is passed as params to second run
Okay, given that each service is supposed to be very small and revolve around one thing this makes sense.

>  "REALLY REALLY need state across methods" then it means that something is wrong with whole idea
Not necessarily, having no state is more suitable for functional paradigm and we're just mixing them up. 

> Why do you think that classes are meant to have state?
I don't, but since we're using Ruby modules seem to be better suited for stateless logic, wouldn't you agree?
Currently our services are just vessels for the functions, that's it.

Reply:
There is difference between arguments and dependencies.
Dependencies are available through whole service object and are the same in each moment(that's why we need to keep it as a class, not module) and arguments are available only in scope of 1 method and they're passed(sometimes totally different) for each method in class

Reply:
```
def initialize(params = {})
  @order = params[:order]
end

attr_reader :order
```
Order is available through the whole object and it'll be always the same unless there are methods with side effects and you call them. How is that different from what you've proposed?

I don't really like the fact that we're mixing two architectural styles and justifying it with DI considering there are alternate implementations.

Reply:
That's definitely the case. `Order will be always the same UNLESS`. I don't want any "unless" and that's why I want to enforce one format of service objects. That's how my solution differs from putting everything in initializer. I'm eliminating ways to do wrong things like basing on state which can be mutated at any time by anyone.

Using instance variables usually leads to terrible code(I've seen it many times across different companies) and I want to avoid it.

I don't understand part about "mixing two architectural styles".  Can you explain?

Reply:
>  I'm eliminating ways to do wrong things like basing on state which can be mutated at any time by anyone

But you can't prevent somebody from mutating the variables in their own services, right? They still can use something like 
```
run RetrieveOrder.new
run MyCustomService.new

class MyCustomService
  def call(order:)
    order.update!(amount: 10)
  end
end
```
Your approach is simply preventing mutation in the initializer but you can't avoid `IF` and `UNLESS` completely, wouldn't you agree?

> Using instance variables usually leads to terrible code

I disagree, I've seen a lot of beautiful code that leverages instance variables & memoization in PORO objects, you just have to use them correctly without introduction of side effects.

Okay, I think it's pointless to discuss it any further. 
Let's try doing it your way, we can always change things once some problems occur.